### PR TITLE
[SplitLogicalObjFifos] Generalize splitting analysis and enable pass in pipeline

### PR DESCRIPTION
**Pull request by jtuyls**
Generalizes the logical objectFifo splitting analysis to find a source, target and objectFifo splitting dimensions. Enables to pass to demonstrate correct functionality on all tests and especially batch matmul.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp:70]: auto iter = std::find_if(memrefShape.begin(), memrefShape.end(),
                         [](int64_t size) { return size > 1; });
auto objFifoSplitDim = std::distance(memrefShape.begin(), iter);

I guess an edge case is when they're all 1, not sure what the correct behaviour is then

**Review by newling — APPROVED** [review]
Small comments. It sounds like this code will be rewritten soon though, so approving if you want to consider these suggestions as future work

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp]: Thanks, using this now. The edge case of all ones should not split, so added a check for this at this place.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp]: Yeah, with partial_sum, this is now resolved as well.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp]: Thanks for the pointer! Using this now.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.h
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.h]: I think having a default `splitFactor` is ok. I changed it to use `std::optional` though now and added documentation.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp:122]: Looks like these codes are mostly duplicated by just switching source/target. Might be good to simplify these.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp:722]: what's the meaning of these comments `// i * `?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp:556]: I think you can remove this condition if you have a separate check for offsetIndices.empty()?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp:607]: Complete this comment.

**Review by yzhang93 — COMMENTED** [review]
Some comments after the PR is merged, these can be considered to address in a follow-up. 

Also it would be good to add more tests regarding different situations of splitDim and splitFactor.

**Pull request merged**
